cmd/bosun/sched/sched.go

```diff
@@ -49,10 +52,29 @@ type Schedule struct {
 	db            *bolt.DB
 }
 
-func (s *Schedule) TimeLock(t miniprofiler.Timer) {
-	t.Step("lock", func(t miniprofiler.Timer) {
-		s.Lock()
-	})
+func init() {
+	metadata.AddMetricMeta(
+		"bosun.schedule.lock.wait", metadata.Counter, metadata.MilliSecond,
+		"The length of time spent waiting for the schedule lock.")
+	metadata.AddMetricMeta(
+		"bosun.schedule.lock.held", metadata.Counter, metadata.MilliSecond,
+		"The length of time spent while holding the schedule lock.")
+}
+
+func (s *Schedule) Lock(method string) {
+	start := time.Now()
+	s.mutex.Lock()
+	s.mutexAquired = time.Now()
+	s.mutexWaitTime = s.mutexAquired.Sub(start) / time.Millisecond // remember this so we don't have to call put until we leave the critical section.
+}
+
+func (s *Schedule) Unlock() {
+	holder := s.mutexHolder
+	start := s.mutexAquired
+	waitTime := s.mutexWaitTime
+	s.mutex.Unlock()
+	collect.Put("bosun.schedule.lock.wait", opentsdb.TagSet{"caller": holder}, waitTime)
```

Comment:
Instead of only recording the last event, how about making this a counter, so we just increase the total time. We would then also add another metric that is the count of locks acquired by a caller. This would allow us to collect data about the average lock hold time for each caller and find calls that are outside the standard deviation. It would also prevent missing any of the slow calls, which Put may certainly do.